pkg/controller/storagecluster/update.go

```diff
@@ -0,0 +1,604 @@
+func (c *Controller) controlledHistories(
+	cluster *corev1alpha1.StorageCluster,
+) ([]*apps.ControllerRevision, error) {
+	// List all histories to include those that don't match the selector anymore
+	// but have a ControllerRef pointing to the controller.
+	historyList := &apps.ControllerRevisionList{}
+	err := c.client.List(context.TODO(), &client.ListOptions{Namespace: cluster.Namespace}, historyList)
+	if err != nil {
+		return nil, err
+	}
+
+	histories := make([]*apps.ControllerRevision, 0)
+	for _, history := range historyList.Items {
+		historyCopy := history.DeepCopy()
+		histories = append(histories, historyCopy)
+	}
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing histories
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (metav1.Object, error) {
+		fresh, err := k8s.Instance().GetStorageCluster(cluster.Name, cluster.Namespace)
+		if err != nil {
+			return nil, err
+		}
+		if fresh.UID != cluster.UID {
+			return nil, fmt.Errorf("original StorageCluster %v/%v is gone, got uid %v, wanted %v",
+				cluster.Namespace, cluster.Name, fresh.UID, cluster.UID)
+		}
+		return fresh, nil
+	})
+
+	selector := c.storageClusterSelectorLabels(cluster)
+	// Use ControllerRefManager to adopt/orphan as needed. Histories that don't match the
+	// labels but are owned by this storage cluster are released (disowned). Histories that
+	// match the labels and do not have ref to this storage cluster are owned by it.
+	cm := controller.NewControllerRevisionControllerRefManager(
+		c.crControl,
+		cluster,
+		labels.SelectorFromSet(selector),
+		controllerKind,
+		canAdoptFunc,
+	)
+	return cm.ClaimControllerRevisions(histories)
+}
+
+func (c *Controller) snapshot(
+	cluster *corev1alpha1.StorageCluster,
+	revision int64,
+) (*apps.ControllerRevision, error) {
+	patch, err := getPatch(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	hash := computeHash(&cluster.Spec, cluster.Status.CollisionCount)
+	name := historyName(cluster.Name, hash)
+	historyLabels := c.storageClusterSelectorLabels(cluster)
+	historyLabels[defaultStorageClusterUniqueLabelKey] = hash
+
+	history := &apps.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            name,
+			Namespace:       cluster.Namespace,
+			Labels:          historyLabels,
+			Annotations:     cluster.Annotations,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(cluster, controllerKind)},
+		},
+		Data:     runtime.RawExtension{Raw: patch},
+		Revision: revision,
+	}
+
+	err = c.client.Create(context.TODO(), history)
+	if outerErr := err; errors.IsAlreadyExists(outerErr) {
+		// TODO: Is it okay to get from cache?
+		existedHistory := &apps.ControllerRevision{}
+		getErr := c.client.Get(
+			context.TODO(),
+			types.NamespacedName{
+				Name:      name,
+				Namespace: cluster.Namespace,
+			},
+			existedHistory,
+		)
+		if getErr != nil {
+			return nil, getErr
```

Comment:
may be you want to wrap the getErr with why we were doing Get in the first place.
I am assuming this will show up when someone does a get status of Storage Cluster.

Reply:
same for others in this if condition